Batching on `.extract_faces` to improve performance and utilize GPU in full

galthran-wq opened the pull request:
## Tickets

https://github.com/serengil/deepface/pull/1433
https://github.com/serengil/deepface/issues/1101
https://github.com/serengil/deepface/issues/1434

### What has been done

With this PR, `.extract_faces` is able to accept a list of images

## How to test

```shell
make lint && make test
```

Benchmarking on detecting 50 faces:
![image](https://github.com/user-attachments/assets/258cab80-08ca-4e3c-b0e4-a19f933a83b6)

For yolov11n, batch size 20 is 59.27% faster than batch size 1.
For yolov11s, batch size 20 is 29.00% faster than batch size 1.
For yolov11m, batch size 20 is 31.73% faster than batch size 1.
For yolov8, batch size 20 is 12.68% faster than batch size 1.

skyler14 commented:
Do you have a branch in your fork that currently combines all the optimizations you've submitted. I'd like to start using them while the approval process is going

whats been the total speedup you've been able to see

galthran-wq commented:
I do. You can check
https://github.com/galthran-wq/deepface/tree/master-enhanced 

it combines these two PRs with some other small modifications:
- `.represent` uses batched detector inference (here https://github.com/serengil/deepface/pull/1433 it only does batched embedding, because batched detection is not yet implemented)
- `.represent` returns a list of list of dicts, if a batch of images is passed. This is neccessary to be able to recover to which images the resulting faces correspond to. It might be a good idea to include this change in the PR as well. You can check the test in the fork.

Not all of the detectors currently (both in this PR and in the fork) implement batching. In particular, YOLO does. I've found it to be optimal in terms of performance and inference speed. The only problem is installing both torch and tensorflow with GPU, but I've managed to somehow do that.


All in all, with the combination of `yolov11m` and `Facenet`, both using GPU, and batch size 100 (the largest I could fit in 4090) I am seeing aroung 15x speed boost, but that is highly dependent on the input images, the GPU (especially memory size). I've also had a quick peek and it seems like the performance on the CPU is improved as well.


@serengil FYI I would be happy to contribute the aforementioned modifications if we have progress on the PRs.

serengil commented:
I will review this PR this week i hope

serengil commented:
Seems this breaks the unit tests. Must be sorted.

galthran-wq commented:
should be good now

serengil commented:
Nope, still failing.

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,50 @@ def test_different_detectors():
+    "ssd"
+])
+def test_batch_extract_faces(detector_backend):
+    img_paths = [
```
do you mind to add couple.jpg here? it has 2 faces. so, result should be list of 4 items and for the one coming from couples.jpg should have many items.

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,50 @@ def test_different_detectors():
 
 
+@pytest.mark.parametrize("detector_backend", [
+    "opencv",
```
please add mtcnn, retinaface and all non-optional ones

serengil commented on deepface/models/face_detection/RetinaFace.py:
```diff
@@ -13,64 +13,72 @@ class RetinaFaceClient(Detector):
-            resp.append(facial_area)
-
-        return resp
+        if isinstance(img, np.ndarray):
```
what if img is (2, 224, 224, 3) sized numpy array? i mean many images in single numpy array.

serengil commented:
You implemented OpenCv, Ssd, Yolo, MtCnn and RetinaFace to accept list inputs

What if I send list to YuNet, MediaPipe, FastMtCnn, Dlib or CenterFace?

I assume an exception will be thrown, but users should see a meaningful message.

galthran-wq commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+        )
+
+
+def test_batch_extract_faces_single_image():
```
FYI this might be not the expected behaviour

serengil replied:
what do you mean?

serengil replied:
what do you mean here?

galthran-wq replied:
I just wanted to draw your attention to the fact that previously, if a single image is passed in a list, it was interpreted as not batched input, so that a list of faces is returned (not a list of lists).
It is not the case in the current state ofthe code, it was fixed.
So this test currently just sanity checks this special case

serengil commented:
@galthran-wq you are pushing too many things, would you please inform me when it is ready.

galthran-wq commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+def test_batch_extract_faces(detector_backend):
+    detector_backend_to_rtol = {
+        "opencv": 0.1,
+        "mtcnn": 0.2,
```
turns out batching has different impact on the result for different backends. It is mostly the same (for almost all the models, the relative error is <1%.
mtcnn and opencv suffer a bit more, not quite sure why (which features are different exactly)

serengil replied:
we may consider to throw exception for those detectors as they are not suitable for batch maybe

galthran-wq commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+    "retinaface",
+    "yunet",
+    "centerface",
+    # optional
```
this might be a good idea to include requirements_additional.txt in the test action so that we could test those as well

serengil replied:
nope, because unit tests will take 1 hour then

galthran-wq commented:
> @galthran-wq you are pushing too many things, would you please inform me when it is ready.

it is ready now, im usually pushing when it's done

I've just fixed what you've noted:
- added psedu-batching (for loop) for other models
- support for np array
- changed detect_faces to return a list of list on batched input
- couple more tests

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+        len(imgs_objs_batch) == 4 and 
+        all(isinstance(obj, list) for obj in imgs_objs_batch)
+    )
+    assert all(
```
please add comment, the last one has many faces, others have just one face

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
         logger.info(f"✅ extract_faces for {detector} backend test is done")
 
 
+@pytest.mark.parametrize("detector_backend", [
```
please add an unit test case for

```python
img1 = cv2.imread(img1_path)
img2 = cv2.imread(img2_path)

img = np.stack([img1, img2])

assert len(img.shape) == 4 # Check dimension.
assert img.shape[0] == 2 # Check batch size.
```

what if image is batch in numpy format?

serengil commented on deepface/models/face_detection/OpenCv.py:
```diff
@@ -29,55 +29,63 @@ def build_model(self):
-        resp = []
-
-        detected_face = None
+        if isinstance(img, np.ndarray):
```
I recommend you to call detect_faces itself it if it is list. this can also be done in Detector.py or detection.py.

in that way, we will not change all detectors

serengil commented:
is this ready? i can still see my comments not resolved.

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+    ]
+    
+    # Extract faces in batch
+    imgs_objs_batch = DeepFace.extract_faces(
```
please check the return type

when it is batch input, then return type should be List[List[Dict[str, Any]]

similarly, put return type check for non batch input test, too - in test_different_detectors case

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+        len(imgs_objs_batch[i]) == 1 
+        for i in range(len(imgs_objs_batch[:-1]))
+    )
+    assert len(imgs_objs_batch[-1]) == 2
```
it is hard to understand why the last one should have 2 and others should have 1

recommend to store expected_num_faces list - https://github.com/serengil/deepface/blob/master/tests/test_represent.py#L212

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +79,144 @@ def test_different_detectors():
+    imgs_batch = np.stack(imgs, axis=0)
+
+    # extract faces in batch of numpy arrays
+    imgs_objs_batch = DeepFace.extract_faces(
```
check the size of imgs_objs_batch - it should be 4 - len(img_paths)

also last item should have 2 items

galthran-wq commented:
To summarize what's changed:

- I've added comments and additional checks to the tests.
- I've made batching on `opencv` and `mtcnn` optional (due to the above issue). To enforce batching a user can set `ENABLE_OPENCV_BATCH_DETECTION` (or `ENABLE_MTCNN_BATCH_DETECTION`) to `true`.

Unfortunately, this didn't fix the batch extraction case on `opencv` -- the problem is that is *occasionally* fails (it seems that the predictions might be different). Note that it has nothing to do with batching, because it is disabled by default. We might add a separate issue and test to reproduce this.

- i have fixed the special case of a single image in a list input (batch of size 1). It now indeed returns a list with a single element -- the list of detected faces in that image.
- those detectors that do no implement batching all had repeating logic in `detect_faces`. I have moved this logic to the default implementation in `Detector`. Now, those detectors only need to implement `_process_single_image`, and batching would be supported by inheritance.
- If a detector implements batching, then it overrides `detect_faces` with this logic, just as before.

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +83,169 @@ def test_different_detectors():
+    )
+
+    # Check that the batch extraction returned the expected number of face lists
+    assert len(imgs_objs_batch) == 4
```
instead of it is equal to 4, check it is equal to len(img_paths)

otherwise if i add a new image to img_path, tests will fail

serengil commented on deepface/modules/detection.py:
```diff
@@ -80,135 +82,158 @@ def extract_faces(
+    batched_input = (
+        (
+            isinstance(img_path, np.ndarray) and
+            img_path.ndim == 4
```
`isinstance(img_path, np.ndarray) and img_path.ndim == 4 and img_path.shape[0] == 1` -> this is not a batched input, right?

serengil commented on deepface/models/Detector.py:
```diff
@@ -72,3 +49,56 @@ class DetectedFace:
+                instead of observer.
+        """
+        raise NotImplementedError(
+            "Subclasses that do not implement batch detection must implement this method"
```
it is not clear that which subclass is giving this error. IMO, we should print the detector name in the exception message.

serengil commented:
I don't like this approach:

```
those detectors that do no implement batching all had repeating logic in detect_faces. I have moved this logic to the default implementation in Detector. Now, those detectors only need to implement _process_single_image, and batching would be supported by inheritance.
If a detector implements batching, then it overrides detect_faces with this logic, just as before.
```

all detectors should have `detect_faces` instead of `_process_single_image`, and you should raise an error in that detector if it is not supported.

galthran-wq commented:
Changes:

1. reverted `_process_single_image`
2. disabled option to use batch mode for mtcnn and opencv

so now they just use a for loop to process all the images, if a batch is passed. A user gets a warning that there is no actual batching happening.

Do you still think it is a good idea to raise an error?

serengil commented on deepface/models/face_detection/MtCnn.py:
```diff
@@ -1,5 +1,6 @@
 # built-in dependencies
-from typing import List
+import logging
```
we are not using logging module directly anywhere. use this instead:

```python
from deepface.commons.logger import Logger
logger = Logger()
```

serengil commented on deepface/models/face_detection/MtCnn.py:
```diff
@@ -8,6 +9,8 @@
 # project dependencies
 from deepface.models.Detector import Detector, FacialAreaRegion
 
+logger = logging.getLogger(__name__)
```
logger = Logger()

serengil commented on deepface/models/face_detection/OpenCv.py:
```diff
@@ -1,6 +1,7 @@
 import os
-from typing import Any, List
+from typing import Any, List, Union
+import logging
```
same here.

serengil commented on deepface/models/face_detection/RetinaFace.py:
```diff
@@ -13,64 +13,75 @@ class RetinaFaceClient(Detector):
-            resp.append(facial_area)
-
-        return resp
+        is_batched_input = isinstance(img, list)
```
you are writing same code in every detector

```python
        is_batched_input = isinstance(img, list)
        if not is_batched_input:
            imgs = [img]
        else:
            imgs = img
```

IMO, this should be done in the parent where those detectors are being called. In that way, we will not have repeated code.

serengil commented on deepface/models/face_detection/RetinaFace.py:
```diff
@@ -13,64 +13,75 @@ class RetinaFaceClient(Detector):
+
+            batch_results.append(resp)
+
+        if not is_batched_input:
```
same here, we can do this in parent and avoid repeated code.

```python
        if not is_batched_input:
            return batch_results[0]
        return batch_results
```

serengil commented on deepface/models/face_detection/MtCnn.py:
```diff
@@ -16,45 +19,71 @@ class MtCnnClient(Detector):
-
-                resp.append(facial_area)
+        img_rgb = [img[:, :, ::-1] for img in img]
+        if self.supports_batch_detection:
```
self.supports_batch_detection is returning false always. so, get rid of _supports_batch_detection.

have just this

```python
detections = [self.model.detect_faces(single_img) for single_img in img_rgb]
```

serengil commented on deepface/models/face_detection/OpenCv.py:
```diff
@@ -29,55 +32,72 @@ def build_model(self):
         return detector
 
-    def detect_faces(self, img: np.ndarray) -> List[FacialAreaRegion]:
+    def _supports_batch_detection(self) -> bool:
```
no need _supports_batch_detection, it is returning false always

serengil commented on deepface/models/face_detection/OpenCv.py:
```diff
@@ -29,55 +32,72 @@ def build_model(self):
-        detected_face = None
+        if isinstance(img, np.ndarray):
+            imgs = [img]
+        elif self.supports_batch_detection:
```
this condition is always false, get rid of it

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +83,169 @@ def test_different_detectors():
 
 
+@pytest.mark.parametrize("detector_backend", [
+    # "opencv",
```
activate opencv please, it is default detector. important to have tests.

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +83,169 @@ def test_different_detectors():
+    expected_num_faces = [1, 1, 1, 2]
+
+    # load images as numpy arrays
+    imgs_batch = np.stack(imgs, axis=0)
```
please add this control

`assert imgs_batch.ndim == 4 and imgs_batch.shape[0] == 4`

serengil commented on tests/test_extract_faces.py:
```diff
@@ -79,6 +83,169 @@ def test_different_detectors():
+        img_path=[img_path],
+        align=True,
+    )
+    assert len(imgs_objs_batch) == 1 and isinstance(imgs_objs_batch[0], list)
```
please check this

assert len(imgs_objs_batch[0]) == 2

serengil commented:
> so now they just use a for loop to process all the images, if a batch is passed. A user gets a warning that there is no actual batching happening.

for loop is okay, don't raise error

serengil commented:
As I understand, you built a for loop in each detector. Is that really a batch?

What if we build a for loop in the parent of detectors, and don't change detectors?

serengil commented:
I mean batch is really running for yolo

We can create a method extract faces from batch in parent class of detectors, and only customize this for yolo.

In parent class, we basically build a for loop and call detector

In that way, detectors will not change

serengil commented:
this has been inactive for a long time, will close it if no objection
